DOC Kymatio documentation 0.2.0 [1]

MuawizChaudhary opened the pull request:
As discussed in #432 , it's better to split the pull request into smaller pull request. This pull request changes the documentation in anticipation of 0.2.0. 

Questions: should the readme be included in this pull request or some other separate one? 

Do we want to create a new branch that is just a copy of master, merge all 0.2.0 changes into that one, then merge into master?

Also see #441 .

Recommended to be merged 1st.

janden commented on doc/source/index.rst:
```diff
@@ -33,26 +31,32 @@
 * integrates wavelet scattering in a deep learning architecture,
-* supports 1-D, 2-D, and 3-D scattering transforms
+* supports 1-D, 2-D, and 3-D scattering transforms,
 * differentiable transforms for applications in generative modeling, reconstruction and more!
```
Should end in `, and`, since it's part of a list.

janden commented on doc/source/installation.rst:
```diff
@@ -1,7 +1,9 @@
 
-There are two ways to install kymatio: via the `pip` package management system, or directly from source.
+There are two ways to install kymatio: via the `pip` package management system, or directly from source. As of Kymatio 0.2.0,
+the PyTorch frontend is the default frontend. In Kymatio 0.3.0, this will be
```
Just 0.2 and 0.3 is more straightforward.

janden replied:
This still says 0.2.0 and 0.3.0 as far as I can see.

janden replied:
For some reason, the change was made, but isn't showing up in this diff.

janden commented on doc/source/quickstart.rst:
```diff
@@ -14,30 +14,27 @@ In the Python interpreter, you may then call::
 
-    import torch
+    import numpy as np
     from kymatio import Scattering2D
```
`from kymatio.numpy import Scattering2D`

janden replied:
Still says `from kymatio`.

janden replied:
Never mind. Can't figure out why GH interface isn't showing the latest version here.

janden commented on doc/source/userguide.rst:
```diff
@@ -114,15 +114,34 @@ size:
+Frontends
+=========
+
+You can choose a frontend via the `frontend=` arguments available in all the Scattering constructors. By default, `pytorch`
```
Backticks around Scattering. Also, I think it's just `torch` for the name of the frontend.

janden commented on doc/source/userguide.rst:
```diff
@@ -114,15 +114,34 @@ size:
+To call this frontend, simply do::
+
+    import numpy as np
+    from kymatio import Scattering2D
```
`kymatio.numpy`

janden commented on doc/source/userguide.rst:
```diff
@@ -114,15 +114,34 @@ size:
     from kymatio import Scattering2D
 
-    scattering = Scattering2D(J=2, shape=(32, 32))
+    scattering = Scattering2D(J=2, shape=(32, 32), frontend='torch')
```
`kymatio.torch`

janden commented on doc/source/userguide.rst:
```diff
@@ -114,15 +114,34 @@ size:
 
 When initialized, a scattering transform object is stored on the CPU::
 
     import torch
```
Needed?

MuawizChaudhary replied:
Also seems like its needed in this example as we use torch functions later on.

edouardoyallon replied:
Yeah- I think it's a good practice, even for us, to explicitely `import torch` .. I find it clearer

janden replied:
> Also seems like its needed in this example as we use torch functions later on.

Ok. Didn't see that. Makes sense.

janden commented on doc/source/userguide.rst:
```diff
@@ -152,34 +171,32 @@ scattering transform object back to the CPU by calling::
+
+To call this frontend, simply do::
+
+    import tensorflow as tf
```
Needed?

janden commented on doc/source/userguide.rst:
```diff
@@ -152,34 +171,32 @@ scattering transform object back to the CPU by calling::
+    import tensorflow as tf
+    from kymatio import Scattering2D
+
+    scattering = Scattering2D(J=2, shape=(32, 32), frontend='tensorflow')
```
`kymatio.tensorflow`

janden commented on doc/source/userguide.rst:
```diff
@@ -192,70 +209,79 @@ Currently, two backends exist:
+    from kymatio import Scattering2D
+    from kymatio.scattering2d.backend.torch_skcuda_backend import backend
+
+    scattering = Scattering2D(J=2, shape=(32, 32), frontend='torch', backend=backend)
```
Again, this needs to be
```
from kymatio.torch import Scattering2D

scattering = Scattering2D(J=2, shape=(32, 32), backend='torch_skcuda')
```

edouardoyallon commented:
ok w/ everything

lostanlen commented on README.md:
```diff
@@ -14,7 +14,7 @@ Scattering transforms are translation-invariant signal representations implement
 * supports 1-D, 2-D, and 3-D wavelets,
 * integrates wavelet scattering in a deep learning architecture, and
-* runs seamlessly on CPU and GPU hardware.
+* runs seamlessly on CPU and GPU hardware, with many deep learning API.
```
too vague

MuawizChaudhary replied:
Thanks for the comment! I believe your looking at an outdated version of the README however.  I replaced this phrase in a recent commit with "with major deep learning APIs (PyTorch and TensorFlow)." per Jokiam Anden's comments. 

Resolve if that sounds fine with you.

edouardoyallon replied:
I'm ok with it.

edouardoyallon commented:
@janden merge?

janden commented on doc/source/installation.rst:
```diff
@@ -33,9 +31,10 @@ Finally, the package is installed by running::
+running::
 
-    pip install -r requirements_optional_cuda.txt
+    pip install scikit-cuda cupy
```
Why can't we have this in an *optional* requirements file again? Hardcoding dependencies is too brittle.

MuawizChaudhary replied:
@edouardoyallon

MuawizChaudhary replied:
@edouardoyallon

edouardoyallon replied:
@eickenberg is the one who knows.

janden commented on README.md:
```diff
@@ -27,7 +28,7 @@ The resort to PyTorch tensors as inputs to Kymatio allows the programmer to back
 Each of these algorithms is written in a high-level imperative paradigm, making it portable to any Python library for array operations as long as it enables complex-valued linear algebra and a fast Fourier transform (FFT).
 
-Currently, there are two available backends, PyTorch (CPU and GPU) and scikit-cuda (GPU only).
+Currently, there are four available frontend-backends, PyTorch (CPU and GPU), PyTorch+SciKit-cuda (GPU only), TensorFlow (CPU and GPU) and NumPy (CPU).
```
`scikit-cuda` (all lowercase).

MuawizChaudhary commented on doc/source/userguide.rst:
```diff
@@ -152,34 +169,30 @@ scattering transform object back to the CPU by calling::
+
+To call this frontend, simply do::
+
+    from kymatio.tensorflowimport Scattering2D
```
Good catch

MuawizChaudhary commented on doc/source/userguide.rst:
```diff
@@ -114,13 +114,30 @@ size:
+Frontends
+=========
+
+You can choose a frontend via the `frontend=` arguments available in all the `Scattering` constructors. By default, `pytorch`
```
I wrote some stuff, do look over it and let me know what you think.

edouardoyallon replied:
Maybe we could leave it as it is from now and move to the next PR, and improving later this part of the code?

janden replied:
Looks good.

lostanlen commented on README.md:
```diff
@@ -27,7 +28,7 @@ The resort to PyTorch tensors as inputs to Kymatio allows the programmer to back
 Each of these algorithms is written in a high-level imperative paradigm, making it portable to any Python library for array operations as long as it enables complex-valued linear algebra and a fast Fourier transform (FFT).
 
-Currently, there are two available backends, PyTorch (CPU and GPU) and scikit-cuda (GPU only).
+Currently, there are four available frontend-backends, PyTorch (CPU and GPU), PyTorch+scikit-cuda (GPU only), TensorFlow (CPU and GPU) and NumPy (CPU).
```
unclear

lostanlen replied:
unclear what "frontend-backends" means

lostanlen commented on README.md:
```diff
@@ -14,7 +14,8 @@ Scattering transforms are translation-invariant signal representations implement
 * supports 1-D, 2-D, and 3-D wavelets,
 * integrates wavelet scattering in a deep learning architecture, and
-* runs seamlessly on CPU and GPU hardware.
+* runs seamlessly on CPU and GPU hardware, with major deep learning APIs
```
remove parens

lostanlen commented on README.md:
```diff
@@ -52,18 +53,13 @@ Andreux M., Angles T., Exarchakis G., Leonarduzzi R., Rochette G., Thiry L., Zar
-
-Once PyTorch is installed, you may install the latest version of Kymatio using the package manager `pip`, which will automatically download Kymatio from the Python Package Index (PyPI):
+We strongly recommend running Kymatio in an Anaconda environment, because this simplifies the installation of other
+APIs. You may install the latest version of Kymatio using the package manager `pip`, which will automatically download
```
APIs -> dependencies

lostanlen commented on doc/source/quickstart.rst:
```diff
@@ -14,30 +14,27 @@ In the Python interpreter, you may then call::
-The performance of the scattering transform depends on the specific backend
-that's in use. For more information on switching backends to improve
-performance, see the :ref:`backend-story` section.
+The performance of the scattering transform depends on the frontend as well as the specific backend that's in use. For
```
avoid apostrophe

janden replied:
I don't know. Do we need to be that formal?

lostanlen commented on doc/source/userguide.rst:
```diff
@@ -152,34 +174,30 @@ scattering transform object back to the CPU by calling::
 
 The scattering transform implementation in Kymatio is structured around a
-flexible backend framework. These backends encapsulate the most computationally
+flexible frontend-backend framework. The backends encapsulate the most computationally
```
unclear

lostanlen replied:
unclear

edouardoyallon commented:
maybe we could merge, and later modify? I highly doubt this will be the final version and I think we spend way too much time on this.

lostanlen commented on README.md:
```diff
@@ -14,7 +14,8 @@ Scattering transforms are translation-invariant signal representations implement
 * integrates wavelet scattering in a deep learning architecture, and
-* runs seamlessly on CPU and GPU hardware.
+* runs seamlessly on CPU and GPU hardware, with major deep learning APIs
+  (PyTorch and TensorFlow).
```
remove parens

janden commented on doc/source/userguide.rst:
```diff
@@ -117,9 +117,14 @@ size:
 
-You can choose a frontend via the `frontend=` arguments available in all the `Scattering` constructors. By default, `pytorch`
+You can choose a frontend via the import `from kymatio.frontend import
+Scattering2D`, where `frontend` is either `numpy`, `tensorflow`, or `pytorch`.
```
It's just `'torch'`, right?

janden commented on doc/source/userguide.rst:
```diff
@@ -117,9 +117,14 @@ size:
+You can choose a frontend via the import `from kymatio.frontend import
+Scattering2D`, where `frontend` is either `numpy`, `tensorflow`, or `pytorch`.
+
+Additionally, you can choose a frontend via the `frontend=` arguments available in all the `Scattering` constructors. By default, `pytorch`
```
Again, `'torch'`.

janden commented:
This all looks good. Just that last `'pytorch'` → `'torch'` and we should be good.

> I think we spend way too much time on this.

Well in some sense fleshing out the documentation is fleshing out the specification for what we want the interface to look like, so it's good to get it right the first time.

MuawizChaudhary commented:
I think @lostanlen makes some good points that the documentation is unclear at some parts. I added in some sentences and removed one more two. Please look over these changes and merge if everything looks good!

edouardoyallon commented:
Despite the jenkins' bug, merging is good for me. What should we do? Squash and merge? @janden @eickenberg

janden commented:
> Despite the jenkins' bug, merging is good for me. 

Looks like Jenkins has downgraded NumPy to 1.11 (and Python to 3.5) for some reason. Tests pass when I run them locally, so I think it's fine. @dylex is looking into it.

> What should we do? Squash and merge?

Yeah I'll squash this into a few commits (since it's not all doc changes) and merge. Are we all good here, @lostanlen?

janden commented:
> Tests pass when I run them locally, so I think it's fine.

This has now been resolved in #448. Once we get that merged, we'll rebase this on top and tests should pass.

janden commented:
Rebasing.

@lostanlen, need your go-ahead for merge.

edouardoyallon commented:
Great, how do we merge it?(seems there was a rebase somewhere)
